NDTensors/src/lib/Sectors/src/category_product.jl

```diff
@@ -34,78 +51,231 @@ function Base.show(io::IO, s::CategoryProduct)
+  g = reduce(×, fused)
+  # convention: keep unsorted blocklabels as produced by F order loops in ×
+  type_fixed = recover_category_product_type(T, g)
+  return type_fixed
+end
+
+function recover_category_product_type(T::Type, g0::AbstractGradedUnitRange)
+  new_labels = recover_category_product_type.(T, blocklabels(g0))
+  new_blocklengths = labelled.(unlabel.(blocklengths(g0)), new_labels)
+  return gradedrange(new_blocklengths)
+end
+
+function recover_category_product_type(T::Type, c::AbstractCategory)
+  return recover_category_product_type(T, CategoryProduct(c))
+end
+
+function recover_category_product_type(T::Type, c::CategoryProduct)
+  return recover_category_product_type(T, categories(c))
+end
+
+function recover_category_product_type(T::Type{<:CategoryProduct}, cats)
+  return recover_category_product_type(categories_type(T), cats)
+end
+
+function recover_category_product_type(T::Type, cats)
+  return CategoryProduct(T(cats))
+end
+
+# =================================  Cartesian Product  ====================================
 ×(c1::AbstractCategory, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 function ×(p1::CategoryProduct, p2::CategoryProduct)
   return CategoryProduct(categories_product(categories(p1), categories(p2)))
 end
 
-categories_product(l1::NamedTuple, l2::NamedTuple) = union_keys(l1, l2)
-
-categories_product(l1::Tuple, l2::Tuple) = (l1..., l2...)
-
+×(a, g::AbstractUnitRange) = ×(to_gradedrange(a), g)
+×(g::AbstractUnitRange, b) = ×(g, to_gradedrange(b))
 ×(nt1::NamedTuple, nt2::NamedTuple) = ×(CategoryProduct(nt1), CategoryProduct(nt2))
 ×(c1::NamedTuple, c2::AbstractCategory) = ×(CategoryProduct(c1), CategoryProduct(c2))
 ×(c1::AbstractCategory, c2::NamedTuple) = ×(CategoryProduct(c1), CategoryProduct(c2))
 
-#
-# Dictionary-like implementation
-#
+function ×(l1::LabelledInteger, l2::LabelledInteger)
+  c3 = label(l1) × label(l2)
+  m3 = unlabel(l1) * unlabel(l2)
+  return labelled(m3, c3)
+end
+
+function ×(g1::AbstractUnitRange, g2::AbstractUnitRange)
+  v = map(
+    ((l1, l2),) -> l1 × l2,
+    Iterators.flatten((Iterators.product(blocklengths(g1), blocklengths(g2)),),),
+  )
+  return gradedrange(v)
+end
+
+# ====================================  Fusion rules  ======================================
+# generic case: fusion returns a GradedAxes, even for fusion with Empty
+function fusion_rule(::NotAbelianStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return to_gradedrange(categories_fusion_rule(categories(s1), categories(s2)))
+end
+
+# Abelian case: fusion returns CategoryProduct
+function fusion_rule(::AbelianStyle, s1::CategoryProduct, s2::CategoryProduct)
+  return categories_fusion_rule(categories(s1), categories(s2))
+end
+
+# Empty case
+function fusion_rule(::AbelianStyle, ::TrivialSector, ::TrivialSector)
+  return CategoryProduct(())
+end
+
+# TrivialSectorStyle acts as trivial on any AbstractCategory, not just CategoryProduct
```

Comment:
This comment looks like a mistake, there isn't any `TrivialSectorStyle` from what I can see.